### PR TITLE
Fix/login providers

### LINKED ISSUE #6570: Cannot get any authentication provider to work: redirect_uri_mismatch
**Describe the bug**
I followed the documentation, I tried using the react-login-front-end-app, no matter how hard I try, I always run into the same redirect_uri_mismatch error.

**Steps to reproduce the behavior**
1. Select any provider [in your strapi admin panel](http://localhost:1337/admin/plugins/users-permissions/providers), let's go with GitHub, since it seems to be the simplest and also has a [video tutorial from Strapi](https://www.youtube.com/watch?v=s_vQqQ4tpHs) (tried following it step by step).
2. Fill in the required fields as [described in the docs](https://strapi.io/documentation/v3.x/plugins/users-permissions.html#providers); also follow the other steps described in the docs
3. Go to http://localhost:1337/connect/github or click on the Github button inside the React login frontend boilerplate app

**Expected behavior**
Successful authentication

**Screenshots**
See in **Additional context**

**System**
- Node.js version: 12.16.1
- NPM version: 6.14.5
- Strapi version: 3.0.1
- Database: Postgres 12.3
- Operating system: Windows 10 Pro x64

**Additional context**
Seems to be a problem with how strapi is sending the redirect_uri to the Oauth providers. Shouldn't it be something like https://760bbe10e952.ngrok.io/connect/github/callback ?
image
image

I also noticed that, as opposed to the tutorial video, redirect URL field in the provider config doesn't get automatically populated, could it be a frontend issue?
image

If I try any other provider (tried facebook & google as well), I also get an URL related error.

### LINKED ISSUE #7206: Impossible to use a new custom authentication provider
**Describe the bug**

Impossible to use a new custom authentication provider.

I have followed the documentation https://strapi.io/documentation/v3.x/plugins/users-permissions.html#adding-a-new-provider-to-your-project

The code of my custom authentication provider is available in the repository https://github.com/madarche/my-project with each step in a separated commit (https://github.com/madarche/my-project/commits/master) :

1. Initial commit
2. Copy from `node_modules`
3. Add missing file from the documentation
4. Add new custom OIDC provider

Related discussions, but with no replies:
* https://github.com/strapi/strapi/discussions/6455
* https://github.com/strapi/strapi/discussions/6584

Related issue:
* https://github.com/strapi/strapi/issues/891 with the associated
  repository https://github.com/abdonrd/strapi-demo but obsolete because it's 2 years old

The fact that there are people having problems creating such a new authentication provider, with no online example of a working new custom authentication provider can be seen as an issue in itself. Understanding *Purest* REST API Configuration is hard. But maybe the issue is not only the need for more doc, maybe is there a bug behind?

**Steps to reproduce the behavior**

1. Setup the new authentication provider as specified by the doc
2. See that the new provider is successfully available in http://localhost:1337/admin
3. Enable the new provider
4. Then with a browser visit URL http://localhost:1337/connect/myoidc
5. See browser being redirected to
http://localhost:1337/auth/myoidc/callback?error=Grant%3A%20missing%20or%20misconfigured%20provider
with the following JSON response:

    ```
    statusCode	400
    error	"Bad Request"
    message
    message	"No access_token."
    data
    message	"No access_token."
    ```

**Expected behavior**

The authentication provider should be contacted and the browser should be redirected to the provider's login form.

**Screenshots**

[The new provider is successfully available in http://localhost:1337/admin](https://github.com/madarche/my-project/blob/master/Screenshot_2020-07-26-Strapi-Roles_Permissions.png)

**Code snippets**

https://github.com/madarche/my-project

**System**

- Node.js version: 12.18.2
- NPM version: 6.14.5
- Strapi version: 3.1.1
- Database: SQLite
- Operating system: Debian GNU-Linux 9 (stretch)

**Additional context**

Before creating this issue I've been trying to use a custom authentication provider with different versions of Strapi: 3.0.x, 3.1.0. 3.1.1 without success.

My custom authentication provider is an *OpenID Connect* provider, but that has no impact whatsoever since this *OpenID Connect* provider is never contacted by Strapi.

I've been able to successfully access the GitHub login form without problem when I enabled the bundled `GitHub` provider.

### PR DESCRIPTION
**Pull request by petersg83**
⚠ need to merge #7054 before
fix #7206
fix #6570

Should be be merged at the same time as: https://github.com/strapi/strapi-examples/pull/164

- Provider doc is updated
- use of "redirect_url" is simplified in front

NB:
I tested all providers. They all work except:
- twitter (the twitter user email was not given in the response api, so strapi can't match it with a strapi user)
- instagram (the api is most likely outdated)
- vk (api is outdated, min version is 5.21)

Thoses bugs need to be fixed, but not in this PR.

### CODE REVIEW

File: packages/strapi-plugin-users-permissions/admin/src/components/PopUpForm/index.js [hunk 1]
```diff
@@ -73,40 +79,16 @@ class PopUpForm extends React.Component {
         return `${strapi.backendURL}/connect/vk/callback`;
       case 'twitch':
         return `${strapi.backendURL}/connect/twitch/callback`;
-      default: {
-        const value = get(this.props.values, 'callback', '');
-
-        return startsWith(value, 'http') ? value : `${strapi.backendURL}${value}`;
-      }
+      default:
+        return `${strapi.backendURL}/connect/\${provider}/callback`;
```
Comment (soupette) [inline packages/strapi-plugin-users-permissions/admin/src/components/PopUpForm/index.js:83]: Why the backslash ?

File: docs/v3.x/plugins/users-permissions.md [hunk 2]
```diff
@@ -156,53 +156,95 @@ axios
...
-Before setting up a provider, you'll need to install the `ngrok` package globally to work with providers that don't allow `localhost` redirect URIs.
+For better understanding, you may find as follows the description of the login flow. To simplify the explanation, we used `github` as the provider but it works the same for the other providers.
 
-In the following examples, the client app will be the [react login examples app](https://github.com/strapi/strapi-examples/tree/master/login-react). It will be running on `http://localhost:3000`.
+#### Understanding the login flow
+
+Let's say that strapi's backend is located at: strapi.website.fr.
+Let's say that your app frontend is located at: website.fr.
+
+1. The user goes on your frontend app (`https://website.fr`) and click on your button `connect with Github`.
```
Comment (lauriejim) [inline docs/v3.x/plugins/users-permissions.md:166]: Can you please replace all
```diff
- website.fr
+ website.com
```

File: docs/v3.x/plugins/users-permissions.md [hunk 3]
```diff
@@ -156,53 +156,95 @@ axios
...
+4. Once done, Github redirects the tab to `https://strapi.website.fr/connect/github/callback?code=abcdef` that calls the backend
+5. The backend uses the given `code` to get from Github an `access_token` that can be used for a period of time to make authorized requests to Github to get the user info (the email of the user of example).
+6. Then, the backend redirects the tab to the url of your choice with the param `access_token` (example: `http://website.fr/connect/github/redirect?access_token=eyfvg`)
+7. The frontend (`http://website.fr/connect/github/redirect`) calls the backend with `https://strapi.website.fr/auth/github/callback?access_token=eyfvg` that returns the strapi user profile with its `jwt`. <br> (Under the hood, the backend asks Github for the user's profile and a match is done on Github user's email address and Strapi user's email address)
+8. The frontend now possesses the user's `jwt`, with means the user is connected and the frontend can make authenticated requests to the backend!
+
+An example of a frontend app that handles this flow can be found here: [react login example app](https://github.com/strapi/strapi-examples/tree/master/login-react).
```
Comment (lauriejim) [inline docs/v3.x/plugins/users-permissions.md:175]: Do we have to replace it with your new app?

File: packages/strapi-plugin-users-permissions/admin/src/components/PopUpForm/index.js (hunk 1 above)
Comment (petersg83) [reply packages/strapi-plugin-users-permissions/admin/src/components/PopUpForm/index.js:83]: Thanks, I'll remove it :)

File: docs/v3.x/plugins/users-permissions.md (hunk 3 above)
Comment (petersg83) [reply docs/v3.x/plugins/users-permissions.md:175]: if https://github.com/strapi/strapi-examples/pull/164 is merged, then the url should be the same

Comment (lauriejim) [reply docs/v3.x/plugins/users-permissions.md:175]: Okay

File: packages/strapi-plugin-users-permissions/services/Providers.js [hunk 4]
```diff
@@ -392,3 +393,11 @@ const getProfile = async (provider, query, callback) => {
       break;
   }
 };
+
+const buildRedirectUri = (provider = '') =>
+  `${getAbsoluteServerUrl(strapi.config)}/connect/${provider}/callback`;
```
Comment (dastein1) [inline packages/strapi-plugin-users-permissions/services/Providers.js:398]: Could somebody explain why this remains pointing to _/connect/..._ . Shouldn't that point to _/auth/..._ as this is the route configured [here](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-users-permissions/config/routes.json#L235)

Comment (petersg83) [reply packages/strapi-plugin-users-permissions/services/Providers.js:398]: We use the lib `grant` and it generatse itself the redirect_uri (as it is it itself that needs the answer from the provider).
In the admin panel, we generate the `redirect_uri` the same way grant does.
We can see it in its doc here: https://create-react-app.dev/docs/adding-custom-environment-variables/ at the line `redirect_uri`.

**Comment by lauriejim** [comment]
Is this PR ready to merge?

**Comment by petersg83** [comment]
> Is this PR ready to merge?

I'm waiting for a review from @alexandrebodin

**Comment by lauriejim** [comment]
Can you please target the `documentation` branch.

**Comment by alexandrebodin** [comment]
@lauriejim this needs to be merged into master as there is code changes :)

**Comment by derrickmehaffy** [comment]
Thank you for this @petersg83 works great

**Comment by veerjainATgmail** [comment]
Hi @petersg83 , thanks for these changes.

For the GItHub provider.... just following strapi-examples/login-react from the replaceLoginApp branch is sufficient or 
need to wait for any strapi side changes as well?

Thanks,
mj.

**Comment by veerjainATgmail** [comment]
Just tested it with the Github provider and replaceLoginApp branch login-React exaple.
Works perfect.! Thanks. The login-React example is simpler now. 

Notes:
If someone is facing github error "Authorization callback URL mismatch error", that means url is not mentioned in the config/server.js file. Pasting the snapshot for the working case.

http://localhost:1337/admin/plugins/users-permissions/providers
https://github.com/settings/applications
and, config/server.js - strapi server code.

image

image

image

**Pull request merged**